### PR TITLE
HyperNova IVC

### PR DESCRIPTION
**Pull request by sjudson**
Completing https://github.com/nexus-xyz/nexus-zkvm/issues/76, this PR implements HyperNova (CycleFold) IVC, including benchmarking.

### CODE REVIEW

File: nova-benches/benches/recursive-snark.rs [hunk 1]
```diff
@@ -39,63 +48,130 @@ criterion_group! {
...
+                .verify(black_box(&pp), black_box(NUM_WARMUP_STEPS))
+                .unwrap();
         });
+    });
+
+    group.finish();
+}
+
+#[cfg(feature = "spartan")]
+fn hypernova(c: &mut Criterion, num_cons_in_step_circuit: usize) {
```
Comment (sjudson) [inline nova-benches/benches/recursive-snark.rs:109]: Reminder to myself to factor out into a separate file.

File: nova/src/ccs/mod.rs [hunk 2]
```diff
@@ -72,6 +72,21 @@ pub struct CCSShape<G: CurveGroup> {
     pub cSs: Vec<(G::ScalarField, Vec<usize>)>,
 }
 
+impl<G: CurveGroup> fmt::Display for CCSShape<G> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "CCSShape {{ num_constraints: {}, num_vars: {}, num_io: {}, num_matrices: {}, num_multisets: {}, max_cardinality: {}, Ms: [_, {}], cSs: [_, {}] }}",
```
Comment (slumber) [inline nova/src/ccs/mod.rs:77]: nit `A: [_, {N}]` in R1CS means that A is a matrix with `N` non-zero elements. Here it's `Ms: [_, {number of M matrices}]`

Comment (sjudson) [reply nova/src/ccs/mod.rs:77]: I mean this is mostly a gut check anyway, the most relevant size information is captured in the first three parameters

File: nova/src/lib.rs [hunk 3]
```diff
@@ -21,9 +21,10 @@ pub mod commitment;
 pub mod r1cs;
 
 pub use circuits::{
+    hypernova::{self}, // uses same StepCircuit trait as Nova
```
Comment (slumber) [inline nova/src/lib.rs:24]: How's clippy happy though

File: nova/src/circuits/hypernova/sequential/mod.rs [hunk 4]
```diff
@@ -0,0 +1,590 @@
...
+        let mut rng = rand_chacha::ChaCha20Rng::from_seed(seed);
+
+        let max_poly_vars = ((shape.num_vars.max(shape.num_constraints) - 1)
+            .checked_ilog2()
+            .unwrap_or(0)
+            + 1) as usize;
+
+        let srs = C1::setup(max_poly_vars, b"hypernova_seq_primary_curve", &mut rng)
+            .map_err(|_| cyclefold::Error::PolyCommitmentSetup)?;
+
```
[the remark this answers is not in the document]
Comment (sjudson) [reply nova/src/circuits/hypernova/sequential/mod.rs:119]: Ah I see, take in the SRS as SetupAux, will do.

File: nova/src/provider/zeromorph.rs [hunk 5]
```diff
@@ -0,0 +1,108 @@
+use std::marker::PhantomData;
```
[the remark this answers is not in the document]
Comment (sjudson) [reply nova/src/provider/zeromorph.rs:1]: I could have, but it has its own logging that I didn't want to stomp on. Plus integrating the SRS from SetupAux stuff will add concrete functionality that makes sense here instead of Spartan (as in the pcd/compression code).

File: nova/src/circuits/hypernova/sequential/mod.rs [hunk 6]
```diff
@@ -0,0 +1,590 @@
...
+    pub fn verify(
+        &self,
+        params: &PublicParams<G1, G2, C1, C2, RO, SC>,
+        num_steps: usize,
+    ) -> Result<(), cyclefold::Error> {
+        let _span = tracing::debug_span!(target: LOG_TARGET, "verify", %num_steps).entered();
+
+        const NOT_SATISFIED_ERROR: cyclefold::Error =
+            cyclefold::Error::Ccs(crate::ccs::Error::NotSatisfied);
```
[the remark this answers is not in the document]
Comment (sjudson) [reply nova/src/circuits/hypernova/sequential/mod.rs:400]: IIRC clippy didn't like it for some reason, but I can take another look + potentially just tell clippy to allow.

File: nova/src/provider/zeromorph.rs [hunk 7]
```diff
@@ -91,7 +94,21 @@ where
         )
         .entered();
 
-        SpartanZM::setup(max_poly_vars, label, rng)
```
Comment (slumber) [inline nova/src/provider/zeromorph.rs]: Why is this change needed?

In practice it's not going to be used as SRS requires trusted setup.

Comment (sjudson) [reply nova/src/provider/zeromorph.rs]: Just trying to make test setup as low touch as possible

Comment (sjudson) [reply nova/src/provider/zeromorph.rs]: I guess I can push this elsewhere though, will change shortly

Comment (slumber) [reply nova/src/provider/zeromorph.rs]: Okay I mean why does this code **need** to be somewhere.
This kind of matters for Pedersen bases, for tests you can use test_rng and it's good enough without additional hashing and seeds.

Comment (sjudson) [reply nova/src/provider/zeromorph.rs]: I guess that's true, okay I can strip it down to the `test_rng`.

File: nova/src/utils.rs [hunk 8]
```diff
@@ -35,6 +35,13 @@ pub fn iter_bits_le(bytes: &[u8]) -> impl Iterator<Item = bool> + '_ {
         .flat_map(|byte| (0..8).map(move |bit| ((1 << bit) & byte) != 0))
 }
 
+#[macro_export]
+macro_rules! safe_log {
+    ($x:expr) => {
+        ($x - 1).checked_ilog2().unwrap_or(0) + 1
+    };
+}
+
```
Comment (slumber) [inline nova/src/utils.rs:44]: 1. It's good to add a doc-comment explaining what it does: computes the number of bits needed to represent numbers in range
[0; x)
For example, given an array of N elements you need this many bits to represent each index.
2. It's not really safe:
```rust
let i = std::hint::black_box(0u64);
println!("{}", safe_log!(i));
```
prints 64 in release. I don't think we ever use it with zeros (it's asserted in most places), but it's either the name or the underflow check.

**Review by slumber — APPROVED** [review]

File: nova/src/utils.rs (hunk 8 above)
Comment (sjudson) [reply nova/src/utils.rs]: Done.

File: nova/src/utils.rs [hunk 9]
```diff
@@ -35,10 +35,17 @@ pub fn iter_bits_le(bytes: &[u8]) -> impl Iterator<Item = bool> + '_ {
         .flat_map(|byte| (0..8).map(move |bit| ((1 << bit) & byte) != 0))
 }
 
+/// An implementation of base_2 log that will never fail, instead returning 0 for all values smaller than 1.
+/// It is 'safe' in the sense that it will not panic, unlike just calling `.unwrap()` on a `.checked_ilog2`.
+/// Usually used to compute the number of bits needed to index a set or a vector.
```
Comment (slumber) [inline nova/src/utils.rs:40]: It should mention that it returns ceil of log2 then, "safety" isn't the only difference from std ilog2, with a special case of `safe_log!(1) = 1`

File: nova/src/utils.rs [hunk 10]
```diff
@@ -35,10 +35,17 @@ pub fn iter_bits_le(bytes: &[u8]) -> impl Iterator<Item = bool> + '_ {
...
 #[macro_export]
 macro_rules! safe_log {
     ($x:expr) => {
-        ($x - 1).checked_ilog2().unwrap_or(0) + 1
+        match $x {
+            x if x <= 1 => 0,
```
[the remark this answers is not in the document]
Comment (slumber) [reply nova/src/utils.rs:45]: Although seeing that the code of mle.rs changed I'm not sure that we care about x = 1

File: nova/src/utils.rs [hunk 11]
```diff
@@ -35,15 +35,21 @@ pub fn iter_bits_le(bytes: &[u8]) -> impl Iterator<Item = bool> + '_ {
...
+///   log_2(x) = y for x > 1
+///              1 for x = 1
+///              0 otherwise
+///
 /// It is 'safe' in the sense that it will not panic, unlike just calling `.unwrap()` on a `.checked_ilog2`.
-/// Usually used to compute the number of bits needed to index a set or a vector.
+/// Used to compute the number of bits needed to index a set or a vector.
 #[macro_export]
-macro_rules! safe_log {
+macro_rules! safe_loglike {
```
Comment (slumber) [inline nova/src/utils.rs:48]: `ceil_log2` maybe?

Comment (sjudson) [reply nova/src/utils.rs:48]: It's not quite `ceil` though, and neither is it `clamped` because `log_2 = 0` and we're redefining that.
